lightning/src/routing/network_graph.rs

```diff
@@ -1170,6 +1190,74 @@ impl NetworkGraph {
 		self.update_channel_intern(msg, None, None::<(&secp256k1::Signature, &Secp256k1<secp256k1::VerifyOnly>)>)
 	}
 
+	/// Override a (directed) channel's individual properties without having a full update
+	/// This method should be used exclusively in conjunction with rapid graph sync, which is intended
+	/// for mobile devices.
+	///
+	/// `short_channel_id`: Channel to be updated
+	///
+	/// `flags`: Channel flags to set. Least significant bit serves as directional indicator
+	///
+	/// `mutations: Vec<ChannelMutationProperty>`: Array of properties to mutate
+	pub fn override_directional_channel_properties(&self, short_channel_id: u64, flags: u8, timestamp: u32, chain_hash: BlockHash, mutations: ChannelMutationProperties) -> Result<(), LightningError> {
+		let channel_update = {
+			let mut channels = self.channels.read().unwrap();
+			let direction = flags & 1u8;
+			match channels.get(&short_channel_id) {
+				None => return Err(LightningError{err: "Couldn't find channel for update".to_owned(), action: ErrorAction::IgnoreError}),
+				Some(channel) => {
+					let directional_info = if direction == 0 {
+						channel.one_to_two.as_ref()
+					} else {
+						channel.two_to_one.as_ref()
+					};
+					if let Some(directional_info) = directional_info {
+
+						let htlc_maximum_msat = if let Some(htlc_maximum_msat) = directional_info.htlc_maximum_msat {
+							OptionalField::Present(htlc_maximum_msat)
+						} else {
+							OptionalField::Absent
+						};
+
+						let mut synthetic_update = msgs::UnsignedChannelUpdate {
+							chain_hash,
+							short_channel_id,
+							timestamp,
+							flags,
+							cltv_expiry_delta: directional_info.cltv_expiry_delta,
+							htlc_minimum_msat: directional_info.htlc_minimum_msat,
+							htlc_maximum_msat,
+							fee_base_msat: directional_info.fees.base_msat,
+							fee_proportional_millionths: directional_info.fees.proportional_millionths,
+							excess_data: vec![]
+						};
+
+						// apply mutations where provided
+						if let Some(cltv_expiry_delta) = mutations.cltv_expiry_delta {
+							synthetic_update.cltv_expiry_delta = cltv_expiry_delta;
+						}
+						if let Some(htlc_minimum_msat) = mutations.htlc_minimum_msat {
+							synthetic_update.htlc_minimum_msat = htlc_minimum_msat;
+						}
+						if let Some(fee_base_msat) = mutations.fee_base_msat {
+							synthetic_update.fee_base_msat = fee_base_msat;
+						}
+						if let Some(fee_proportional_millionths) = mutations.fee_proportional_millionths {
+							synthetic_update.fee_proportional_millionths = fee_proportional_millionths;
+						}
+						if let Some(htlc_maximum_msat) = mutations.htlc_maximum_msat {
+							synthetic_update.htlc_maximum_msat = htlc_maximum_msat;
+						}
+						synthetic_update
+					}else{
+						return Err(LightningError{err: "Couldn't find previous directional data for update".to_owned(), action: ErrorAction::IgnoreError});
+					}
+				}
+			}
+		};
+		self.update_channel_unsigned(&channel_update)
```

Comment:
This may need to be atomic with the earlier read. What would happen in the following scenario?
- Form an `UnsignedChannelUpdate` from incremental update using the read-locked `NetworkGraph` above.
- Apply `ChannelUpdate` from failed payment
- Apply earlier formed `UnsignedChannelUpdate`

Seems the incremental update would override one we just received from a failed payment.

Reply:
It shouldn't be a problem, because all the rapid sync updates are backdated a week, so any "true" update from direct gossip should result in this update being discarded.